url8d3c4d17f5_catseye_JaC64_tgz-pJ8-JaC64J8/src/com/dreamfabric/jac64/C64Screen.java

```diff
@@ -0,0 +1,2025 @@
+      break;
+    }
+  }
+
+  // Used to draw background where either border or background should be
+  // painted...
+  private void drawBackground() {
+    int bpos = mpos;
+    int currentBg = borderState > 0 ? borderColor : bgColor; 
+    for (int i = 0; i < 8; i++) {
+      mem[bpos++] = currentBg;
+    }
+  }
+
+  /**
+   * <code>drawGraphics</code> - draw the VIC graphics (text/bitmap)
+   * Note that sprites are not drawn here... (yet?)
+   *
+   *
+   * @param mpos an <code>int</code> value representing position to
+   * draw the graphics from (already fixed with hscroll)
+   */
+  private final void drawGraphics(int mpos) {
+    if (!gfxVisible || paintBorder || (borderState & 1) == 1) {
+      // We know that display is not enabled, and that mpos is already
+      // at a correct place, except horizScroll...
+      mpos -= horizScroll;
+      int color = (paintBorder || (borderState > 0)) ? borderColor : bgColor;
+      for (int i = mpos, n = mpos + 8; i < n; i++) {
+        mem[i] = color;
+      }
+      // trick to use vmli as a if var even when no gfx.
+      vmli++;
+      return;
+    }
+
+    int collX = (vmli << 3) + horizScroll + SC_XOFFS;
+    
+    // Paint background if first col (should maybe be made later also...)
+    if (vmli == 0) {
+      for (int i = mpos - horizScroll, n = i + 8; i < n; i++) {
+        mem[i] = bgColor;
+      }
+    }
+    
+    int position = 0, data = 0, penColor = 0, bgcol = bgColor;
+
+    if ((control1 & 0x20) == 0) {
+      int tmp;
+      int pcol;
+
+      // This should be in a cache some where...
+      if (multiCol) {
+        multiColor[0] = bgColor;
+        multiColor[1] = cbmcolor[bgCol[1]];
+        multiColor[2] = cbmcolor[bgCol[2]];
+      }
+
+
+      penColor = cbmcolor[pcol = vicColCache[vmli] & 15];
+      if (extended) {
+        position = charMemoryIndex +
+        (((data = vicCharCache[vmli]) & 0x3f) << 3);
+        bgcol = cbmcolor[bgCol[(data >> 6)]];
+      } else {
+        position = charMemoryIndex + (vicCharCache[vmli] << 3);
+      }
+
+      data = memory[position + rc];
+
+      if (multiCol && pcol > 7) {
+        multiColor[3] = cbmcolor[pcol & 7];
+        for (int pix = 0; pix < 8; pix += 2) {
+          tmp = (data >> pix) & 3;
+          mem[mpos + 6 - pix] = mem[mpos + 7 - pix] = multiColor[tmp];
+          // both 00 and 01 => no collission!?
+          // but what about priority?
+          if (tmp > 0x01) {
+            tmp = 256;
+          } else {
+            tmp = 0;
+          }
+          collissionMask[collX + 7 - pix] =
+            collissionMask[collX + 6 - pix] = tmp;
+        }
+      } else {
+        for (int pix = 0; pix < 8; pix++) {
+          if ((data & (1 << pix)) > 0) {
+            mem[mpos + 7 - pix] = penColor;
+            collissionMask[collX + 7 - pix] = 256;
+          } else {
+            mem[mpos + 7 - pix] = bgcol;
+            collissionMask[collX + 7 - pix] = 0;
+          }
+        }
+      }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1577:1587, 1636:1647.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.